rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,226 @@
+constant tensors in to the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+The key and value design of unified constant tensor cache utils:
+
+- Cache key: The unified cache key can be composed of two parts: backend specific
+  key and backend id. The backend id is unique among backends. The backend
+  specific key can be defined in each backends, to achieve more flexibility.
+- Cache value: The unified cache value can be a base buffer class like following,
+  and each backend can inherit this base class and subclass it.
+
+```cpp
+class constant_buffer_t{
+public:
+    virtual size_t get_size() const;
+    virtual ~constant_buffer_t();
+
+private:
+    void *data_;
+    size_t size_;
+    // Memory freeing relies on engine and allocator
+    const dnnl::engine eng_;
+    const allocator_t *alc_;
+};
+```
+
+The constant tensor cache utils may looks like the following. It will be
+templated class which can be specialized and instantiated for each engine kind
+for dedication purpose. The support for concurrent execution is same as the
+existing DNNL backend constant tensor cache, which used the promise-future
+mechanism. That is not changed.
+
+```cpp
+template<class engine_kind>
+class unified_constant_tensor_cache_t {
+public:
+    using key_t = size_t;
+    using cached_t = std::shared_ptr<constant_buffer_t>;
+    using value_t = std::shared_future<cached_t>;
+    using promise_t = std::promise<cached_t>;
+
+    unified_constant_tensor_cache_t();
+    ~unified_constant_tensor_cache_t();
+
+    status_t set_capacity(size_t capacity);
+    size_t get_capacity();
+
+    /// @backend_id:
+    /// @backend_specific_key:
+    /// @size: The size the constant tensor
+    /// @value: A promised future constant tensor. If cache miss, the thread
+    /// which firstly query the constant tensor will promise to compute it in
+    /// the future, then other threads who also query the same constant tensor
+    /// will wait util the first thread ready.
+    /// @return: A valid future of cache hit, and an invalid one if cache miss.
+    value_t get_or_add(size_t backend_id, size_t backend_specific_key,
+            sze_t size, const value_t &value);
+    void remove_if_exist(size_t backend_id, size_t backend_specific_key);
```

Comment:
What if we're removing a tensor cache when it's currently using? What's the suggested way to use this API?

Reply:
> What if we're removing a tensor cache when it's currently using?

This API only removes a tensor from the cache, not destroys the tensor. If the removed tensor is using, the backend users should hold the tensor and make sure it's alive.

> What's the suggested way to use this API?

Backend user could call this API to remove a cached tensor if they know the cached tensor won't be used anymore. For example, if a partition is destroyed, the cached tensors for this partition could most likely be removed.

Reply:
`A backend partition would hold constant cached tensors till the end of partition destruction.`
Since the tensor is removed from the `unified_constant_tensor_cache`, would the `unified_constant_tensor_cache` further allow caching new constant tensors?
If so, imagine the case when all the partitions only get destroyed at the end of the entire execution process, the total amount of backend-cached constant tensors might be much larger than the capacity limit set by the users. Then the specified capacity limit would be meaningless?

Reply:
> A backend partition would hold constant cached tensors till the end of partition destruction. all the partitions only get destroyed at the end of the entire execution process,

For this case, why you remove the tensor from cache?

Reply:
The possible cache and backend co-work style in my mind:
1. when the tensor is in cache and not being used by backend, it's owned by cache
1. when the tensor is in cache and being used by backend, it's co-owned by cache the backend. 
1. when the tensor is removed from cache and being used by backend, it's owned by backend
1. when the tensor is removed from cache and not being used by backend, it will be freed

I think we are talking about the 3rd case. when backend needs to use the tensor, it can query the tensor from cache and hold it temporarily

> A backend partition would hold constant cached tensors till the end of partition destruction.

What's this kind of constant tensor used for, can you give some examples?

Reply:
As was discussed in the meeting, since compiler backend will release backend-cached constant tensor as long as the corresponding `cp` is not under execution. My concern no longer exist. Thanks